core/server/worker/src/main/java/alluxio/worker/block/BlockRWLockPool.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.worker.block;
+
+import alluxio.metrics.MetricKey;
+import alluxio.metrics.MetricsSystem;
+import alluxio.resource.ResourcePool;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Class for managing block read/write locks. After obtaining a lock with
+ * {@link ResourcePool#acquire()}, {@link ResourcePool#release(Object)} must be called when the
+ * thread is done using the client.
+ */
+@ThreadSafe
+public final class BlockRWLockPool extends ResourcePool<ClientRWLock> {
+  private static final Logger LOG = LoggerFactory.getLogger(BlockRWLockPool.class);
+
+  @VisibleForTesting
+  final AtomicBoolean mHasReachedFullCapacity;
+  private final AtomicInteger mRemainingPoolResources;
+
+  /**
+   * Creates a new client read/write lock pool.
+   * @param maxCapacity the max capacity of the pool
+   */
+  public BlockRWLockPool(int maxCapacity) {
+    super(maxCapacity);
+    mHasReachedFullCapacity = new AtomicBoolean(false);
+    mRemainingPoolResources = new AtomicInteger(maxCapacity);
+
+    MetricsSystem.registerGaugeIfAbsent(
+        MetricKey.WORKER_BLOCK_LOCK_POOL_REMAINING_RESOURCES_KEY.toString(),
+        this::getRemainingPoolResources
+    );
+  }
+
+  /**
+   * Get the number of remaining pool resources.
+   * @return the number of remaining resources
+   */
+  @VisibleForTesting
+  public long getRemainingPoolResources() {
+    return mRemainingPoolResources.get();
+  }
+
+  private void maybeLogPoolReachesFullCapacity() {
+    if (this.size() == mMaxCapacity
+        && !mHasReachedFullCapacity.get()
+        && mHasReachedFullCapacity.compareAndSet(false, true)
+    ) {
+      LOG.warn("Block lock manager client RW lock pool exhausted. Capacity size {}", mMaxCapacity);
+    }
+  }
+
+  @Override
+  public ClientRWLock acquire() {
+    return this.acquire(WAIT_INDEFINITELY, null);
+  }
+
+  @Nullable
+  @Override
+  public ClientRWLock acquire(long time, TimeUnit unit) {
+    ClientRWLock ret = super.acquire(time, unit);
+    maybeLogPoolReachesFullCapacity();
```

Comment:
If the full capacity has been reached, I'd actually expect to see some log before getting blocked and wait. So when I see my application stuck somewhere, I can tell from the log why I'm stuck. Therefore printing before and after acquire act slightly differently.

**Print after acquire:**
When the one request takes the last resource, a message will be logged. If no request releases, and some time elapses, and my request comes in and get blocked, it may not be obvious why my request blocks because the previous log can be far ahead in the log.

**Print before acquire:**
If the resource are already out, I can see a message logged right there matching the timestamp of my request. It's not perfect but slightly more obvious.

There's always a gap between the check(log) and `acquire()` so the logs are not 100% precious. Maybe an even better way is to NEVER `acquire()` indefinitely. When the `TimeUnit` is null or the timeout is pretty long, we use a while loop to acquire with smaller timeouts but track the total time elapsed instead. Then the logging is much much more flexible.

Reply:
yup i totally agree with this. 
Actually I have come up with a better solution. For the acquire() function without any args, instead of slicing the lock acquisition process, we can simply do a timed lock acquisition first and give it enough long time, say 1 second. If the lock acquisition failed, we print a log about the thread pool depletion. Then we can turn into the unlimited waiting mode to acquire the lock.